### PR TITLE
torch 3 (next-to-last in series)

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd [hunk 1]
```diff
@@ -0,0 +1,429 @@
+---
+title: "Using torch modules"
+description: > 
+ In this next-to-last installment of a mini-series introducing torch basics, we replace hand-coded matrix operations by modules, considerably simplifying our toy network's code.
```
Comment (batpigandme) [inline _posts/2020-10-07-torch-modules/torch_modules.Rmd:4]:
```suggestion
 In this third installment of our mini-series introducing torch basics, we replace hand-coded matrix operations by modules, considerably simplifying our toy network's code. 
```

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+[Last
+week](https://blogs.rstudio.com/ai/posts/2020-10-01-torch-network-from-scratch),
+we started learning about `torch` basics by coding a simple neural
+network from scratch, making use of just a single of `torch`'s features:
+*tensors*. [On
+Monday](https://blogs.rstudio.com/ai/posts/2020-10-05-torch-network-with-autograd),
+we immensely simplified the task, replacing manual backpropagation with
+*autograd*. Today, we *modularize* the network - in both the habitual
+and a very literal sense: Low-level matrix operations are swapped out
+for `torch` `module`s.
```
Comment (batpigandme) [inline _posts/2020-10-07-torch-modules/torch_modules.Rmd:37]: Since this post will (presumably) be read in the future, I think you're better off referencing previous posts as first and second (or something of the like)—last week and Monday aren't very meaningful without context.

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+
+## Modules
+
+From other frameworks (Keras, say), you may be used to distinguishing
+between *models* and *layers*. In `torch`, both are instances of
+`nn_Module()`, and thus, have some methods in common. For those thinking
+in terms of "models" and "layers", I'm artificially splitting up this
+section into two parts, for clarity of exposition. In reality though,
```
Comment (batpigandme) [inline _posts/2020-10-07-torch-modules/torch_modules.Rmd:45]:
```suggestion
`nn_Module()`, and thus, have some methods in common. For those thinking
in terms of "models" and "layers", I'm artificially splitting up this
section into two parts. In reality though,
```

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+following snippet instantiates a linear layer that expects three-feature
+inputs and returns a single output per observation:
+
+```{r}
+library(torch)
+l <- nn_linear(3, 1)
+```
+
+The module has two parameters, "weight" and "bias". Both have already
+been initialized:
```
Comment (batpigandme) [inline _posts/2020-10-07-torch-modules/torch_modules.Rmd:62]: Automatically by torch or is that what you did in the previous chunk?

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+    -0.0385  0.1412 -0.5436
+    [ CPUFloatType{1,3} ]
+
+    $bias
+    torch_tensor 
+    -0.1950
+    [ CPUFloatType{1} ]
+
+Modules are callable; calling a module executes its `forward()` method,
+which for a linear layer just matrix-multiplies input and weights and
```
Comment (batpigandme) [inline _posts/2020-10-07-torch-modules/torch_modules.Rmd:79]:
```suggestion
which, for a linear layer, matrix-multiplies input and weights and
```

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+     0.1876
+    -0.0930
+     0.7498
+    -0.2332
+    -0.0428
+     0.3849
+    -0.2618
+    [ CPUFloatType{10,1} ]
+
+In addition though, it knows what will need to be done, should ever it
```
Comment (batpigandme) [inline _posts/2020-10-07-torch-modules/torch_modules.Rmd:108]: I'm not sure what "it" is referring to here. If I were to scroll up to trace to the last noun, "it" would refer to `out`, but I'm still not totally sure what that means.

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+be asked to calculate gradients:
+
+```{r}
+out$grad_fn
+```
+
+    AddmmBackward
+
+(You may remember from the preceding post that when we work with tensors
```
Comment (batpigandme) [inline _posts/2020-10-07-torch-modules/torch_modules.Rmd:117]:
```suggestion
You may remember from the preceding post that when we work with tensors
```

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+directly, we need to create them passing `requires_grad = TRUE` to
+achieve this. With modules, `torch` correctly assumes that we'll want to
+perform backpropagation at some point.)
```
Comment (batpigandme) [inline _posts/2020-10-07-torch-modules/torch_modules.Rmd:120]:
```suggestion
perform backpropagation at some point.
```

Comment (batpigandme) [reply _posts/2020-10-07-torch-modules/torch_modules.Rmd:120]: Having a stand-alone parenthetical seems off to me. You could put it as an aside, if you think it's out of place inline with everything else.

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+
+By now though, we haven't called `backward()` yet. Thus, no gradients
```
Comment (batpigandme) [inline _posts/2020-10-07-torch-modules/torch_modules.Rmd:122]:
```suggestion
We haven't called `backward()` yet. Thus, no gradients
```

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd [hunk 10]
```diff
@@ -0,0 +1,429 @@
...
+Of the quantities on the right side, we're interested in the second. We
+need to provide the first one, the way it would look *if really we were
+taking the mean*:
+
+```{r}
+d_avg_d_out <- torch_tensor(10)$`repeat`(10)$unsqueeze(1)$t()
+out$backward(gradient = d_avg_d_out)
+```
+
+Now, `l$weight$grad` and `l$bias$grad` *do* contain the respective
```
Comment (batpigandme) [inline _posts/2020-10-07-torch-modules/torch_modules.Rmd:170]:
```suggestion
Now, `l$weight$grad` and `l$bias$grad` *do* contain their respective
```

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+```
+
+    torch_tensor 
+     1.3410  6.4343 -30.7135
+    [ CPUFloatType{1,3} ]
+    torch_tensor 
+     100
+    [ CPUFloatType{1} ]
+
+Back to the main thread. Besides `nn_linear()` , `torch` provides pretty
```
Comment (batpigandme) [inline _posts/2020-10-07-torch-modules/torch_modules.Rmd:185]:
```suggestion
In addition to `nn_linear()` , `torch` provides pretty
```

**Review by batpigandme — APPROVED** [review]
Nice. As the text gets sparser, you want to be increasingly ruthless getting rid of qualifiers and asides so that it's still easy to find the "meat" of the prose. 👍

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd (hunk 2 above)
Comment (skeydan) [reply _posts/2020-10-07-torch-modules/torch_modules.Rmd:37]: makes sense!

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd (hunk 4 above)
Comment (skeydan) [reply _posts/2020-10-07-torch-modules/torch_modules.Rmd:62]: right, making that clearer...

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd (hunk 8 above)
Comment (skeydan) [reply _posts/2020-10-07-torch-modules/torch_modules.Rmd:120]: oh yeah, you're spotting everything :-)

Thinking about it, I even changed this to a paragraph starting "Note the difference..." :-)

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd (hunk 10 above)
Comment (skeydan) [reply _posts/2020-10-07-torch-modules/torch_modules.Rmd:170]: yeah, right... actually i'm probably being overly clear here as often ... `do contain gradients` should suffice...

File: _posts/2020-10-07-torch-modules/torch_modules.Rmd (hunk 6 above)
Comment (skeydan) [reply _posts/2020-10-07-torch-modules/torch_modules.Rmd:108]: yeah, changing ....

**Comment by skeydan** [comment]
thanks a lot Mara! Whatever the topic, you always spot when something's left a bit unclear :-))
